Fix: correct user_id validation logic in Anthropic…

raz-alon opened the pull request:
## Title

Add validation to exclude invalid `user_id` (e.g. email or phone number) from Anthropic metadata

## Relevant issues

Fixes #10106 #11432


## Type
🐛 Bug Fix

## Changes

This PR Fixed the validation check for the `user_id` field in the `_litellm_metadata` dictionary to ensure it does not contain personally identifiable information (PII) such as emails or phone numbers.

## Rationale
Anthropic **requires** that the `user_id` in metadata must be a UUID, hash value, or other opaque, non-identifying identifier. It **must not** contain any personally identifiable information such as names, email addresses, or phone numbers. Including such data can lead to the API rejecting the request with a `400 Bad Request error`. [View](https://docs.anthropic.com/en/api/messages#body-metadata-user-id)

This update ensures that if an invalid user_id is provided (e.g. an email), it will simply be excluded from the metadata, preventing avoidable request failures.

raz-alon committed:
Fix Anthropic user ID validation across both chat and completion endpoints

- Fixed completion endpoint to validate user IDs before setting metadata
- Added proper import of _valid_user_id function in completion transformation
- Prevents BadRequestError: 'user_id appears to contain an email address'
- Added comprehensive test suite with 36 test cases covering:
  - Email validation (both endpoints)
  - Phone number validation
  - Valid user ID acceptance
  - Edge cases (None, non-string, empty strings, whitespace)
  - Metadata validation from litellm_params
  - Error prevention scenarios
  - Integration tests
- All tests pass, ensuring the fix prevents the original BadRequestError
- Aligns with Anthropic's documentation requiring UUID/hash values instead of identifying information

krrishdholakia commented on tests/llm_translation/test_anthropic_completion.py:
```diff
@@ -1344,6 +1345,53 @@ def test_anthropic_prefix_prompt():
     assert response.choices[0].message.content.startswith("Argentina")
 
 
+def test_anthropic_user_id_validation():
```
move inside `test_litellm/`

krrishdholakia commented on tests/llm_translation/test_anthropic_user_id_validation.py:
move mock test inside `test_litellm/` so it runs via the github action

krrishdholakia commented:
bump on this? @raz-alon

raz-alon committed:
Enhance user ID validation in Anthropic transformation

- Updated email regex pattern to improve validation accuracy
- Reintroduced user ID validation tests for both chat and completion endpoints
- Ensured valid user IDs are accepted while invalid ones (emails) are rejected
- All tests pass, confirming the robustness of the validation logic

raz-alon committed:
Add unit tests for Anthropic user ID validation

- Introduced a comprehensive test suite for validating user IDs in Anthropic configurations.
- Tests cover acceptance of valid user IDs, rejection of emails and phone numbers, and handling of edge cases (None, non-string, empty strings, whitespace).
- Ensured integration with litellm completion calls for both chat and text models.
- All tests pass, confirming the robustness of the user ID validation logic.

raz-alon committed:
Remove unused import from Anthropic completion test file

raz-alon committed:
Enhance user ID handling in Anthropic transformation

- Implemented validation for user IDs to ensure only valid IDs are passed to metadata.
- Integrated the _valid_user_id function to filter out emails and phone numbers.
- This change prevents potential errors related to invalid user identification.

raz-alon commented:
> bump on this? @raz-alon

I run the server locally, see everting worked as we want it to be, and also move the tests to the dedicated folder : )
@krrishdholakia

krrishdholakia commented on litellm/llms/anthropic/experimental_pass_through/adapters/transformation.py:
```diff
@@ -349,7 +350,9 @@ def translate_anthropic_to_openai(
             if metadata and "user_id" in metadata:
-                new_kwargs["user"] = metadata["user_id"]
+                # Only pass user_id if it's valid (not an email or phone number)
+                if _valid_user_id(metadata["user_id"]):
```
Why add this on the anthropic to openai bridge?

I don't think openai has the same restrictions

raz-alon commented:
> is this still a valid pr? @raz-alon

Yes I still waiting for this bug to be pushed and fixed...
